cabal-install/src/Distribution/Client/ProjectConfig/Parsec.hs

```diff
@@ -0,0 +1,136 @@
+import Distribution.Client.ProjectConfig.Types (ProjectConfig)
+import Distribution.Fields.ConfVar                   (parseConditionConfVar)
+import Distribution.Fields.ParseResult
+-- AST type
+import Distribution.Fields (Field, readFields', Name (..))
+import Distribution.Fields.LexerMonad                (LexWarning, toPWarnings)
+import Distribution.PackageDescription.Quirks        (patchQuirks)
+import Distribution.Parsec                           (parsec, simpleParsecBS)
+import Distribution.Parsec.Position                  (Position (..), zeroPos)
+import Distribution.Parsec.Warning                   (PWarnType (..))
+import Distribution.Types.CondTree                   (CondTree (..), CondBranch (..))
+import Distribution.Types.ConfVar                    (ConfVar (..))
+import Distribution.Utils.Generic                    (breakMaybe, fromUTF8BS, toUTF8BS, unfoldrM, validateUTF8)
+
+import qualified Data.ByteString                                   as BS
+import qualified Text.Parsec                                       as P
+
+-- TODO 6101 the following is copied from Distribution.PackageDescription.Parsec.parseGenericPackageDescription
+-- - do we need to use patchQuirks also for cabal.project files? if we do, we can extract a
+-- common preprocessing function (patchQuirks, valid UTF8, ) to avoid code duplication
+-- | Preprocess file and start parsing
+parseProjectSkeleton :: BS.ByteString -> ParseResult ProjectConfigSkeleton
+parseProjectSkeleton bs = do
+    case readFields' bs'' of
+        Right (fs, lexWarnings) -> do
+            when patched $
+                parseWarning zeroPos PWTQuirkyCabalFile "Legacy cabal file"
+            parseProjectSkeleton' lexWarnings invalidUtf8 fs
+        Left perr -> parseFatalFailure pos (show perr) where
+            ppos = P.errorPos perr
+            pos  = Position (P.sourceLine ppos) (P.sourceColumn ppos)
+  where
+    (patched, bs') = patchQuirks bs
+    invalidUtf8 = validateUTF8 bs'
+    bs'' = case invalidUtf8 of
+        Nothing -> bs'
+        Just _  -> toUTF8BS (fromUTF8BS bs')
+
+parseProjectSkeleton'
+    :: [LexWarning]
+    -> Maybe Int
+    -> [Field Position]
+    -> ParseResult ProjectConfigSkeleton
+parseProjectSkeleton' lexWarnings utf8WarnPos fs = do
+    parseWarnings (toPWarnings lexWarnings)
+    for_ utf8WarnPos $ \pos ->
+        parseWarning zeroPos PWTUTF $ "UTF8 encoding problem at byte offset " ++ show pos
+
+    -- TODO I am not sure whether we need sectionizeFields for cabal.project parsing
+    -- it is declared in Distribution.PackageDescription.Parsec.sectionizeFields
+    -- and used to convert old-style Cabal files, does this also apply to cabal.project files?
+    -- let (syntax, fs') = sectionizeFields fs
+    -- maybe we can also just use partitionFields
+
+    -- let (fields, sectionFields) = takeFields fs
+    parseCondTree specVer hasElif projectConfigFieldGrammar fs
+    where
+      -- TODO where do we get specVer?
```

Comment:
I am not sure currently how I can get the `CabalSpecVersion` here, but I would be very grateful for a hint.

Reply:
I also don't think that `CabalSpecVersion` is relevant to project files, since it refers to the version of the .cabal file format. I'm not sure if there is a version number for the format of the project file.

Reply:
Thank you very much! I now use `cabalSpecLatest` everywhere. Also I found that it is already used like this somewhere in the legacy projectConfig parser.